### PR TITLE
MIR-borrowck: immutable unique closure upvars can be mutated

### LINKED ISSUE #46023: MIR-borrowck: immutable unique closure upvars can be mutated
MIR borrowck allows for immutable unique closure upvars to be mutated, e.g.
```Rust
fn main() {
    let x = 0;

    (move || {
        x = 1; // mutating immutable local, ok on MIR, bad on AST
    })()
}
```

### PR DESCRIPTION
**Pull request by davidtwco**
Fixes #46023 and #46160 (see [this comment](https://github.com/rust-lang/rust/pull/46236#issuecomment-347204874)).

### CODE REVIEW

**Comment by davidtwco** [comment]
This isn't quite working yet and I'm not quite sure why. I think I've followed the instructions left on the issue correctly.

r? @arielb1 @nikomatsakis

File: src/librustc/mir/mod.rs [hunk 1]
```diff
@@ -544,8 +544,8 @@ impl<'tcx> LocalDecl<'tcx> {
 pub struct UpvarDecl {
     pub debug_name: Name,
 
-    /// If true, the capture is behind a reference.
-    pub by_ref: bool
+    /// If None then the capture is by-value.
```
Comment (arielb1) [inline src/librustc/mir/mod.rs]: rename it to `pub capture_mode: Option<BorrowKind>`?

File: src/librustc_mir/borrow_check.rs [hunk 2]
```diff
@@ -994,7 +994,21 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
...
                                     // Mutably borrowed data is mutable, but only if we have a
                                     // unique path to the `&mut`
-                                    hir::MutMutable => self.is_unique(&proj.base),
+                                    hir::MutMutable => {
+                                        let field_projection = self.is_upvar_field_projection(
+                                            &proj.base);
+
+                                        if let Some(field) = field_projection {
+                                            let decl = &self.mir.upvar_decls[field.index()];
+                                            if let Some(borrow_kind) = decl.by_ref {
```
Comment (arielb1) [inline src/librustc_mir/borrow_check.rs]: nit: `if let Some(BorrowKind::Unique) = decl.by_ref {`

**Comment by arielb1** [comment]
I think you also need to change `check_if_reassignment_to_immutable_state` to use the new `is_mutable` logic, otherwise assignments would not quite work.

Also, I would merge `is_mutable` and `is_unique` and have a single function that returns the value's "strongest" borrow kind. That would allow making the `check_if_reassignment_to_immutable_state` check run only for unique-immutable values and avoid a double error in that case.

**Comment by davidtwco** [comment]
Pushed an alternate approach after speaking with @arielb1 [on Gitter](https://gitter.im/rust-impl-period/WG-compiler-nll?at=5a1b0471232e79134dd461fb) - it isn't working yet but I think I've implemented the `mutability` field in `UpvarDecl` as per our conversation.

File: src/librustc_mir/borrow_check.rs [hunk 3]
```diff
@@ -694,28 +694,13 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
...
+        let lvalue = match self.is_mutable(lvalue) {
+            Ok(()) => return,
+            Err(l) => l,
+        };
+
+        if let Lvalue::Static(_) = lvalue {
+            self.report_assignment_to_static(context, (lvalue, span));
         }
 
         if let Some(mpi) = self.move_path_for_lvalue(lvalue) {
```
Comment (arielb1) [inline src/librustc_mir/borrow_check.rs]: You still need the `move_path_closest_to` change here.

**Comment by arielb1** [comment]
@davidtwco 

BTW, if you're already at this, you could also fix #46160 by marking the closure argument (argument `_1` of a closure) as mutable

**Comment by davidtwco** [comment]
I've not done much testing of what I just pushed, but I've added work from all the instruction provided over the past few days. It includes a rebase onto current master which included #46022 which @arielb1 said I might need.

**Comment by davidtwco** [comment]
Pushed a compiling version of the last changes and rebased ontop of the current master. Wasn't sure what to use for the new parameter to `is_mutable`.

**Comment by eddyb** [comment]
This should have been merged like this, `UpvarDecl` was "supposed" to be debuginfo-only :disappointed:.
But I'm not sure where to put this information - presumably it belongs attached to the closure type.

**Comment by nikomatsakis** [comment]
@eddyb 

> But I'm not sure where to put this information - presumably it belongs attached to the closure type

Hmm. Maybe, but I'm not sure. I guess it depends on how you view the `mut` information. I don't think of it as part of the type system, but rather a kind of lint that sits on top. But I suppose we do issue hard errors and so forth, so it's reasonable to view it as information that belongs in the closure type. It's sort of nice though that the "closure subst" is just types right now, and I'm not sure I want to change that. Another option would be to add some `closure_upvar_info` query that takes a closure def-id and gives further information, though — sort of like the existing `free_vars` calculation?

(On the other hand, maybe it's nice to have all the upvar information collected in one place...)

( \<grumpy\> I still think we should just drop the notion of declaring local variables as `mut` altogether. =)  \</grumpy\> )

**Pull request merged**